src/ZoneServer/World/Storage/Storage.cs

```diff
@@ -0,0 +1,691 @@
+
+			var result = _storageItems.TryGetValue(position, out var item);
+			if (!result)
+				return StorageResult.ItemNotFound;
+
+			return Remove(item, amount, out var removedPosition, out removedAmount);
+		}
+
+		/// <summary>
+		/// Returns true if storage is full.
+		/// </summary>
+		/// <returns></returns>
+		public bool CheckStorageFull()
+		{
+			lock (_syncLock)
+				return _storageItems.Count() >= _maxStorage;
+		}
+
+		/// <summary>
+		/// Returns a dictionary of items in each position
+		/// of storage.
+		/// </summary>
+		/// <returns></returns>
+		public Dictionary<int, Item> GetStorage()
+		{
+			lock (_syncLock)
+				return _storageItems.ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
+		}
+
+		/// <summary>
+		/// Returns specific item, or null if it doesn't exist.
+		/// </summary>
+		/// <param name="objectId">Item object id</param>
+		/// <param name="item">Item found</param>
+		/// <param name="position">Position of item found in storage</param>
+		/// <returns></returns>
+		public StorageResult TryGetItem(long objectId, out Item item, out int position)
+		{
+			item = null;
+			position = -1;
+
+			lock (_syncLock)
+			{
+				var storageItem = _storageItems.FirstOrDefault(v => v.Value.ObjectId == objectId);
+				if (storageItem.Value != null)
+				{
+					item = storageItem.Value;
+					position = storageItem.Key;
+					return StorageResult.Success;
+				}
+			}
+
+			return StorageResult.ItemNotFound;
+		}
+
+		/// <summary>
+		/// Returns specific item in position, or null
+		/// if it doesn't exist.
+		/// </summary>
+		/// <param name="position"></param>
+		/// <returns></returns>
+		public Item TryGetItemAtPosition(int position)
+		{
+			lock (_syncLock)
+				if (_storageItems.TryGetValue(position, out var item))
+					return item;
+
+			return null;
+		}
+
+		/// <summary>
+		/// Returns how many items exist in storage
+		/// </summary>
+		/// <returns></returns>
+		public int GetItemCount()
+		{
+			lock (_syncLock)
+				return _storageItems.Count();
+		}
+
+		/// <summary>
+		/// Returns the max size of this storage
+		/// </summary>
+		/// <returns></returns>
+		public int GetStorageSize()
+		{
+			return _maxStorage;
+		}
+
+		/// <summary>
+		/// Increases max storage size by the given number.
+		/// Storages can never decrease in size.
+		/// </summary>
+		/// <param name="size"></param>
+		public StorageResult AddStorageSize(int size)
+		{
+			if (size <= 0)
+				return StorageResult.InvalidOperation;
+
+			_maxStorage += size;
+
+			return StorageResult.Success;
+		}
+
+		/// <summary>
+		/// Swaps items in two positions.
+		/// Moves item into position2 if empty.
+		/// </summary>
+		/// <param name="position1">Item1 object id</param>
+		/// <param name="position2">Item2 object id</param>
+		/// <returns></returns>
+		public StorageResult Swap(int position1, int position2)
+		{
+			var item1 = this.TryGetItemAtPosition(position1);
+			var item2 = this.TryGetItemAtPosition(position2);
```

Comment:
`TryGet` methods usually return a boolean directly and the actual return value in an out parameter, allowing patterns like this:
```cs
if (foo.TryGetX(out var x))
    // got x, do something with it
```
Deviating from that pattern is potentially confusion, making me think that the method should probably be renamed.